cmd/tether/attach.go

```diff
@@ -140,26 +179,82 @@ func (t *attachServerSSH) start() error {
 	}
 	t.sshConfig.AddHostKey(pkey)
 
-	t.enabled = true
+	t.Enabled()
 	go t.run()
 
 	return nil
 }
 
 // stop is not thread safe with start
-func (t *attachServerSSH) stop() {
+func (t *attachServerSSH) stop() error {
 	defer trace.End(trace.Begin("stop attach server"))
 
-	if t == nil || !t.enabled {
-		return
+	if t == nil {
+		err := fmt.Errorf("attach server is not configured")
+		log.Error(err)
+		return err
 	}
 
-	t.enabled = false
-	conn := t.conn
-	t.conn = nil
+	if !t.IsEnabled() {
+		err := fmt.Errorf("attach server is not enabled")
+		log.Error(err)
+		return err
+	}
+
+	log.Debugf("Setting enabled to false")
+	t.Disabled()
+
+	// This context is used by backchannel and we need to cancel it before
+	// trying to obtain the following lock so that backchannel releases it.
+	log.Debugf("Canceling the context")
+	t.cancel()
 
-	if conn != nil {
-		(*conn).Close()
+	log.Debugf("Acquiring the connection lock")
+	t.conn.Lock()
+	if t.conn.conn != nil {
+		log.Debugf("Closing the connection")
+		(*t.conn.conn).Close()
+		t.conn.conn = nil
+	}
+	t.conn.Unlock()
+	log.Debugf("Released the connection lock")
+
+	return nil
+}
+
+func backchannel(ctx context.Context, conn *net.Conn) error {
+	defer trace.End(trace.Begin("establish tether backchannel"))
+
+	// HACK: currently RawConn dosn't implement timeout so throttle the spinning
+	// it does implement the Timeout methods so the intermediary code can be written
+	// to support it, but they are stub implementation in rawconn impl.
+
+	// This needs to tick *faster* than the ticker in connection.go on the
+	// portlayer side.  The PL sends the first syn and if this isn't waiting,
+	// alignment will take a few rounds (or it may never happen).
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
+
+	ch := make(chan struct{}, 1)
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			(*conn).Close()
+			close(ch)
```

Comment:
`close(ch)` before `(*conn).Close()`
Otherwise HandshakeServer may return and re-enter before `ch` is closed.

Also - **I'm not sure what happens if you call read on a closed rawconn**